rfcs/027-pipeline-intermediate-storage/README.md

```diff
@@ -0,0 +1,147 @@
+
+## Requirements for Intermediate Store
+
+Taking into account the business logic of our services, and given the issues with the current VHS highlighted in [Problem](#problem), there are a number of requirements and desirable features for any alternative store:
+
+1. The ability to insert items into the store with some unique string ID and some specific integer version (set externally to the pipeline from our adapters), with a method to only store data when the version given is greater or equal to what is currently in the store.
+
+2. The ability to retrieve data from the store, where the latest data is always retrieved and the version is returned with the data.
+
+3. Happily stores data in our current range of around 8GB, and scaling beyond this for future proofing.
+
+4. Some form of disk persistence so that we do not need to perform a reindex from scratch in the case of the database going down.
+
+5. Cheaper overall cost than currently, including read / write requests, associated networks costs, and storage costs.
+
+6. Ideally improved performance.
+
+7. Ideally better visibility into the data.
+
+## Store Options
+
+There are a number of different options to use as an intermediate store that I have considered, outlined below.
+
+### Elasticsearch
+
+Pros:
+
+* Used elsewhere in the pipeline so general familiarity with it.
+
+* Kibana would provide a level of visibility into the output at different stages unlike anything else available, and the ability to perform complex queries on our data would likely be an asset to debugging.
+
+* Can access via AWS network address to reduce NAT gateway costs (see [here](https://www.elastic.co/guide/en/cloud/current/ec-getting-started-private.html)).
+
+* Back of the envelope calculations show likely to be much cheaper than current solution.
+
+* Horizontal scalability means quite capable of supporting data at sizes much beyond what we have now.
+
+Cons:
+
+* Requires use of external cloud platform or maintenance of our own Elasticsearch cluster, rather than being a service within AWS.
+
+* Terraform does not control the creation and removal of Elasticsearch indices so might require manual clean up after removing a pipeline.
+
+* No transactions which might prove useful for certain use cases.
+
+Questions:
+
+* Elasticsearch supports [versioning](https://www.elastic.co/blog/elasticsearch-versioning-support), including the ability to index with our own external version numbers. However when using external versioning it only indexes data when the version number is greater than currently stored, whereas our business logic currently expects updates to be processed when versions are greater or equal. Some thought will need to be given whether this might work for us, and if not we would need to implement our own system on top of the Elasticsearch APIs (which may not be particularly resilient to race conditions due to lack of transactions).
```

Comment:
There are read/write consistency guarantees (is this what you mean by transactions?) that may be sufficient for our purposes.

Reply:
By transactions I was referring here to a sequence of queries that can be rolled back upon failure, more the atomicity part of ACID than the consistency.

Reply:
Should probably elaborate on this. The proper meaning of a transaction is a single unit of work performed on the database which is ACID compliant. What I (perhaps incorrectly) was using the word transaction to mean was short hand for a group of operations that all form a single unit of work. This is especially useful when making changes to multiple tables concurrently in a relational database etc.

Elasticsearch does not provide this capability, which I thought perhaps would be useful if we needed to implement our own versioning manually on top of Elasticsearch (such as by making a query to check the version, then making a subsequent insert request). I'm not sure grouped transactions would actually be useful in this case, I haven't given it enough thought, but as @jamesgorrie points out there is a way to get Elasticsearch versioning to behave how we want so it's not actually relevant to this particular use case.